src/controllers/slot.ts

```diff
@@ -12,105 +10,179 @@ function isSlotElement(node: Element): node is HTMLSlotElement {
   return node.nodeName === 'SLOT';
 }
 
+/**
+ * Determines if a given slot string/ref matches the specified slot
+ * @param {HTMLSlotElement} node Node to test against
+ * @param {string|Ref} matcher Ref or selector to match
+ * @return {boolean}
+ */
+function slotMatches(
+  node: HTMLSlotElement,
+  matcher: string | Ref<HTMLSlotElement>
+): boolean {
+  if (matcher === 'default' && !node.hasAttribute('name')) {
+    return true;
+  }
+
+  if (typeof matcher !== 'string') {
+    return matcher.value === node;
+  }
+
+  return matcher === node.name;
+}
+
 type SlotListenerCallback<T extends Element> = (node: T) => void;
 
 interface SlotListener {
+  slot: string | Ref<HTMLSlotElement>;
   selector: string;
   callback: SlotListenerCallback<Element>;
 }
 
 /**
  * Helpers for dealing with slotted content
  */
-export class SlotController extends ElementTrackingController {
+export class SlotController implements ReactiveController {
+  private __host: ReactiveControllerHost & Element;
+  private __hasCalled: boolean = false;
+  private __listeners: Set<SlotListener> = new Set<SlotListener>();
+
   /**
-   * Gets the current ref
-   * @return {Ref}
+   * Gets the slot elements in the current host
+   * @return {HTMLSlotElement[]}
    */
-  public get ref(): Ref {
-    // Casting here to drop the nullability
-    return this._ref as Ref;
+  private get __slots(): HTMLSlotElement[] {
+    if (!this.__host.shadowRoot) {
+      return [];
+    }
+
+    return [...this.__host.shadowRoot.querySelectorAll('slot')];
   }
 
   /**
    * @param {ReactiveControllerHost} host Host to attach to
-   * @param {Ref} ref Ref to observe rather than the host element
    */
-  public constructor(
-    host: ReactiveControllerHost & Element,
-    ref: Ref = createRef()
-  ) {
-    super(host, ref);
+  public constructor(host: ReactiveControllerHost & Element) {
+    this.__host = host;
 
     host.addController(this);
   }
 
-  private __listeners: Set<SlotListener> = new Set<SlotListener>();
+  /**
+   * Determines if a given slot has contents in the host
+   * @param {string|Ref} slot Slot to test against, * for all slots
+   * @param {string} selector Selector to test
+   * @return {boolean}
+   */
+  public has(slot: string | Ref<HTMLSlotElement>, selector?: string): boolean {
+    const slotName = typeof slot === 'string' ? slot : slot.value?.name;
+
+    this.__hasCalled = true;
+
+    if (!slotName) {
+      return false;
+    }
+
+    const slots = this.__slots.filter((slot) => {
+      return (
+        slotName === '*' ||
+        (!slot.hasAttribute('name') && slotName === 'default') ||
+        slot.name === slotName
+      );
+    });
+    const children = slots.reduce<Element[]>((accum, slot) => {
+      for (const child of slot.assignedElements()) {
+        accum.push(child);
+      }
+      return accum;
+    }, []);
```

Comment:
it's your code but `Array.prototype.reduce` always gets a big yikes from me, can you not just do a nested for loop since it's what you're doing but more explicit (IMO)?

Reply:
that's fair. reduce does often lead to confusion/spaghetti-code. i'll change it to a loop